### PR TITLE
[PEA] Mark get_field'ed virtual objects as aliases

### PR DESCRIPTION
**Pull request by caojoshua**
Prior to this patch, we replaced the get_field'ed object with the virtual object saved in the state. The problem is that if we are to support loops, it would be hard to undo this change when merging backedges.

In this patch, we mark the following nodes as aliases:
* LoadP - get_field'ed objects with -XX:-UseCompressedOops
* DecodeN - get_field'ed objects with compressed oops
* CastPP - null-checked objects on DecodeN's because they lack non-null typing data
* CheckCastPP - instanceof checks on CastPP's because they lack exact typing data

Now at any point of execution, there can be many nodes aliasing to the same allocation. When we materialize an object, we need to make sure we replace all nodes that alias to that object with the materialized object.

---

This patch is WIP. We currently have a few regressions.

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:tier1                     2244  2244     0     0
>> jtreg:test/jdk:tier1                               2387  2382     4     1 <<
>> jtreg:test/langtools:tier1                         4430  4417    11     2 <<
   jtreg:test/jaxp:tier1                                 0     0     0     0
   jtreg:test/lib-test:tier1                            31    31     0     0
==============================
TEST FAILURE
```

### CODE REVIEW

File: src/hotspot/share/opto/parse3.cpp [hunk 1]
```diff
@@ -77,10 +77,21 @@ void Parse::do_field_access(bool is_get, bool is_field) {
...
     // Compile-time detect of null-exception?
     if (stopped())  return;
 
+    if (DoPartialEscapeAnalysis) {
+      PEAState& as = jvms()->alloc_state();
+      PartialEscapeAnalysis *pea = PEA();
+      VirtualState* vs = as.as_virtual(pea, original_obj);
+      // The null-checked object node aliases to the same object as the original node.
+      if (vs != nullptr) {
+        pea->add_alias(pea->is_alias(original_obj), obj);
```
Comment (navyxliu) [inline src/hotspot/share/opto/parse3.cpp]: so you want to say that castPP is same as original obj? 
Is it possible that we consider disjoint set over hashmap? 
it's faster to do union and changing 'root' in disjoint set.

Comment (caojoshua) [reply src/hotspot/share/opto/parse3.cpp]: Yes, I think its possible. I think we can revisit our data structures in separate work.

**Comment by caojoshua** [comment]
rebased on top of PEA_beta. no changes in diff

**Comment by caojoshua** [comment]
wont fix

**Pull request closed without merging**